Helper/PaymentResponseHandler.php

```diff
@@ -279,6 +290,61 @@ public function handlePaymentsDetailsResponse(
+                    //Cancel the Authorised Payments
+                    $storeId = $order->getStoreId();
+                    $client = $this->dataHelper->initializeAdyenClient($storeId);
+                    $service = $this->dataHelper->initializeOrdersApi($client);
+                    foreach ($getGiftcardDetails as $giftcardData) {
+                        try {
+                            // Decode JSON response and validate it
+                            $response = json_decode($giftcardData['response'], true);
+                            if (json_last_error() !== JSON_ERROR_NONE || !isset($response['order'])) {
+                                throw new InvalidArgumentException('Invalid giftcard response data');
+                            }
+
+                            // Extract order data and PSPRef
+                            $orderData = $response['order']['orderData'] ?? null;
+                            $pspReference = $response['order']['pspReference'] ?? null;
+
+                            if (!$orderData || !$pspReference) {
+                                throw new InvalidArgumentException('Missing orderData or pspReference in the response');
+                            }
+
+                            // Prepare cancel request
+                            $merchantAccount = $this->configHelper->getAdyenAbstractConfigData("merchant_account", $storeId);
+                            $cancelRequest = [
+                                'order' => [
+                                    'pspReference' => $pspReference,
+                                    'orderData' => $orderData,
+                                ],
+                                'merchantAccount' => $merchantAccount,
+                            ];
+
+                            // Call the cancel service
+                            $cancelResponse = $service->cancelOrder(new CancelOrderRequest($cancelRequest));
+                            $response = $cancelResponse->toArray();
+
+                            if (is_null($response['resultCode'])) {
+                                // In case the result is unknown we log the request and don't update the history
+                                $this->adyenLogger->error(
+                                    "Unexpected result query parameter for cancel order request. Response: " . json_encode($response)
+                                );
+
+                                return false;
+                            }
+                        } catch (\Exception $e) {
+                            // Log the error with relevant information for debugging
+                            $this->adyenLogger->error('Error canceling partial payments', [
+                                'exception' => $e->getMessage(),
+                                'giftcardData' => $giftcardData,
+                            ]);
+                        }
+                    }
+                }
```

Comment:
What would you think about moving this logic to a separate private method? Currently, it decreases the readability and increases the overall complexity of the switch block.